src/complete.c

```diff
@@ -0,0 +1,258 @@
+#include "vctrs.h"
+#include "equal.h"
+
+// -----------------------------------------------------------------------------
+
+static SEXP vec_slice_complete(SEXP x);
+
+// [[ register() ]]
+SEXP vctrs_slice_complete(SEXP x) {
+  return vec_slice_complete(x);
+}
+
+static SEXP vec_locate_complete(SEXP x);
+
+static
+SEXP vec_slice_complete(SEXP x) {
+  SEXP loc = PROTECT(vec_locate_complete(x));
+
+  // Skip `vec_as_location()` in `vec_slice()`
+  SEXP out = vec_slice_impl(x, loc);
+
+  UNPROTECT(1);
+  return out;
+}
+
+// -----------------------------------------------------------------------------
+
+// [[ register() ]]
+SEXP vctrs_locate_complete(SEXP x) {
+  return vec_locate_complete(x);
+}
+
+static SEXP vec_detect_complete(SEXP x);
+
+static
+SEXP vec_locate_complete(SEXP x) {
+  SEXP where = PROTECT(vec_detect_complete(x));
+  SEXP out = r_lgl_which(where, false);
+  UNPROTECT(1);
+  return out;
+}
+
+// -----------------------------------------------------------------------------
+
+static inline void lgl_detect_complete(SEXP x, R_len_t size, int* p_out);
+static inline void int_detect_complete(SEXP x, R_len_t size, int* p_out);
+static inline void dbl_detect_complete(SEXP x, R_len_t size, int* p_out);
+static inline void cpl_detect_complete(SEXP x, R_len_t size, int* p_out);
+static inline void chr_detect_complete(SEXP x, R_len_t size, int* p_out);
+static inline void raw_detect_complete(SEXP x, R_len_t size, int* p_out);
+static inline void list_detect_complete(SEXP x, R_len_t size, int* p_out);
+static void df_detect_complete(SEXP x, R_len_t size, int* p_out);
+
+// [[ register() ]]
+SEXP vctrs_detect_complete(SEXP x) {
+  return vec_detect_complete(x);
+}
+
+static
+SEXP vec_detect_complete(SEXP x) {
+  SEXP proxy = PROTECT(vec_proxy_equal(x));
```

Comment:
We currently define missing values in record vectors to be the proxy rows where all values are missing. In the complete cases though, I suspect we should treat record vectors the same way as atomic vectors, i.e. an element is complete unless it's missing. I.e. if there are some missing values in a row, it's still a complete element.

Reply:
This is tricky because we've worked hard to not special case vctrs-rcrd in the C code. To be completely general we might need `vec_proxy_complete()`, which would default to `vec_proxy_equal()` but we could define:

```r
vec_proxy_complete.vctrs_rcrd <- function(x) {
  proxy <- vector(length = vec_size(x))
  incomplete <- vec_equal_na(x)
  proxy[incomplete] <- NA
  proxy
}
```

Reply:
oh definitely not a new proxy. I was thinking `vec_detect_complete()` to special case vectors that inherit from `data.frame`. For other kinds of vectors it'd just use the equal-na code path and flip the boolean vector.

It's also possible that this function should be `df_detect_complete()`.

Reply:
> It's also possible that this function should be df_detect_complete().

That seems....much simpler. I like it!

Reply:
Note: Consider what happens with a rcrd column in this `df_detect_complete()` function.